web3/_utils/abi.py

```diff
@@ -31,11 +31,12 @@
 )
 
 
-def filter_by_type(_type, contract_abi):
+from typing import Any, Dict, Iterator, List, Optional, Tuple, Type, Union
+def filter_by_type(_type: str, contract_abi: Any) -> Union[List[Dict[str, Union[bool, str]]], List[Dict[str, Union[bool, List[Dict[str, Union[bool, str]]], str]]], List[Union[Dict[str, Union[bool, List[Dict[str, str]], str]], Dict[str, Union[bool, str]]]], List[Dict[str, Union[bool, List[Dict[str, str]], str]]], List[Union[Dict[str, Union[bool, str]], Dict[str, Union[bool, List[Dict[str, Union[bool, str]]], str]]]]]:
```

Comment:
Looks like we need to define some `typing.NewType` and `mypy_extensions.TypedDict` for the overall shape of a contract ABI if we want the typing here to be readable.

Reply:
For sure, there's some even more crazy outputs from monkeytype which makes things very unreadable that I haven't committed as yet.

I'll see how I can start offloading some of these to `typing.NewType` and `mypy_extensions.TypedDict` in the morning.